src/plugins/QuarterMaster.cpp

```diff
@@ -53,87 +53,98 @@ float CalcSupplies::operator()(float sum, const sc2::Unit* unit_) const {
     }
 }
 
-float CalcSupplies::operator()(float sum, const Order& order_) const {
-    switch (order_.ability_id.ToType()) {
-        case sc2::ABILITY_ID::BUILD_NEXUS:
-        case sc2::ABILITY_ID::BUILD_COMMANDCENTER:
-            return sum + 15.0f;
+struct CalcConsumptionRate {
+    float operator()(float sum, const sc2::Unit* unit_) const;
+};
 
-        case sc2::ABILITY_ID::BUILD_PYLON:
-        case sc2::ABILITY_ID::BUILD_SUPPLYDEPOT:
-        case sc2::ABILITY_ID::TRAIN_OVERLORD:
-            return sum + 8.0f;
+float CalcConsumptionRate::operator()(float sum, const sc2::Unit* unit_) const {
+    if (unit_->build_progress != 1.0f)
+        return sum;
+    switch (unit_->unit_type.ToType()) {
+        case sc2::UNIT_TYPEID::PROTOSS_NEXUS:
+        case sc2::UNIT_TYPEID::ZERG_HATCHERY:
+        case sc2::UNIT_TYPEID::TERRAN_COMMANDCENTER:
+        case sc2::UNIT_TYPEID::TERRAN_ORBITALCOMMAND:
+        case sc2::UNIT_TYPEID::TERRAN_PLANETARYFORTRESS:
+            return sum + 1.75f;
+            // SCV = 272 framesToTrain (exactly), TrainRate=1/272
+            // SupplyDepot = 480 framesToTrain, TrainRate = 1/480
+            // SCV's trained while Depot builds = 480/272 = 1.765
+
+        case sc2::UNIT_TYPEID::TERRAN_BARRACKS:
+            return sum + 1.2f;
+            // MarineTrainTime = 18*22.4 = 403.2 (roughly), TrainRate = 1/403
+            // SupplyDepot = 480 framesToTrain, TrainRate = 1/480
+            // Marines trained while Depot builds = 480/403 = 1.191
+
+        case sc2::UNIT_TYPEID::TERRAN_BARRACKSREACTOR:
+            return sum + 2.4f;
+            // Double the rounded 1.2 rate of single Barracks = 2.4
+
+        case sc2::UNIT_TYPEID::TERRAN_BARRACKSTECHLAB:
+            return sum + 2.0f;
+            // MarauderTrainTime = 21*22.4 = 470.4 (roughly), TrainRate = 1/470
+            // SupplyDepot = 480 framesToTrain, TrainRate = 1/480
+            // Marauders trained while Depot builds = 480/470 = 1.02
+            // Marauder is 2 supply, so round to 2.0 overall
+
+            case sc2::UNIT_TYPEID::TERRAN_FACTORY:
+            case sc2::UNIT_TYPEID::TERRAN_FACTORYTECHLAB:
+            case sc2::UNIT_TYPEID::TERRAN_FACTORYREACTOR:
+            case sc2::UNIT_TYPEID::TERRAN_STARPORT:
+            case sc2::UNIT_TYPEID::TERRAN_STARPORTTECHLAB:
+            case sc2::UNIT_TYPEID::TERRAN_STARPORTREACTOR:
+            case sc2::UNIT_TYPEID::ZERG_LAIR:
+            case sc2::UNIT_TYPEID::ZERG_HIVE:
+            case sc2::UNIT_TYPEID::PROTOSS_GATEWAY:
+            case sc2::UNIT_TYPEID::PROTOSS_ROBOTICSFACILITY:
+            case sc2::UNIT_TYPEID::PROTOSS_STARGATE:
+                return sum + 2.0f;  // just add 2.0 for all others for now
```

Comment:
I think this approach is a bit buggy, too, as here we assume that each building will produce something in nearest future, while nothing can be produced, e.g. because bot decided to switch to different unit types or
e.g. for the mech Terran strategy where barracks typically don't produce any units.

This will work for the very simple strategies used in the repo, but not with a more complex setup. 
Maybe combination of the new and old approach could be used?

Reply:
I agree it has that weakness, but I think the more complex solution will only work well once there is some kind of IdealArmyComposition data-structure (should be a separate Enhancement Issue) or CombatUnitSpendRateRatio-vs-EconomySpendRateRatio (possibly a fix for Issue #37, "Reserve resources in the slow queue"). But it is difficult to work on those without simple placeholder solutions that work mostly well. That's why I am trying to create simple fixes for many small problems, so the more complex problems can be attacked in a more-stable/better-working environment. It is hard to test fixes for other Issues otherwise.

Reply:
If possible, I'd like to move forward with this pull until something like Issue #51 "Strategy IdealArmyComposition blocks" enables more precise information to do as you suggest. There's only 1 barracks in Mech play, so that's just 1.2 extra assumed supply-need, and building supply a few seconds early in the mid-game (when there are many production buildings) can be a good thing. As a general rule, building a little extra supply is better than less until there is more precise information to be perfectly 'just-in-time'.

Reply:
Fair enough. Ok, lets go with this approach.